indigo_api/data_migrations/__init__.py

```diff
@@ -94,3 +96,84 @@ def update_namespace(self, xml):
             1)
 
         return xml
+
+
+class AKNeId:
+    """ In a document's XML, translate existing `id` attributes to `eId`
+    and follow the new naming and numbering convention
+    e.g. "sec_2" instead of "section-2"
+    """
+    basic_replacements = {
+        # TODO: make sure all have been added
+        # TODO: add a test for each type
+        "chapter": (r"\bchapter-", "chp_"),
+        "part": (r"\bpart-", "part_"),
+        "subpart": (r"\bsubpart-", "subpart_"),
+        "section": (r"\bsection-", "sec_"),
+        "paragraph": (r"\bparagraph-", "para_"),
+        "article": (r"\barticle-", "article_"),
+        "table": (r"\btable(?=\d)", "table_"),
+        "blockList": (r"\blist(?=\d)", "list_"),
+    }
+
+    def update_xml(self, xml):
+        """ Update the xml,
+            - first by updating the values of `id` attributes as necessary,
+            - and then by replacing all `id`s with `eId`s
+        """
+        doc = Act(xml)
+        self.namespace = doc.namespace
+
+        # basic replacements, e.g. "chapter-" to "chap_"
+        for element, patterns in self.basic_replacements.items():
+            for node in doc.root.xpath(f"//a:{element}", namespaces={"a": self.namespace}):
+                old_id = node.get("id")
+                new_id = re.sub(re.compile(patterns[0]), patterns[1], old_id)
+                node.set("id", new_id)
+                self.update_prefixes(node, old_id, new_id)
+
+        # more complex replacements
+        # TODO: simplify the two below (they're very similar)
+        # subsections
+        for node in doc.root.xpath("//a:subsection", namespaces={"a": self.namespace}):
+            old_id = node.get("id")
+            # e.g. sec_3a.2A.4.1 --> sec_3a.subsec_2A-4-1
+            old_pattern = re.compile(r"^(?P<pre>[^.]+)(?P<post>(\.\d[\dA-Za-z]*)+)")
```

Comment:
Would this be simpler?

```python
pre, post = old_id.split('.', 1)
```

Reply:
Oh, no, that only works if the prefix doesn't have a dot.

Why not re-build the id completely, use the parent element's id as the prefix, and the content of `num` as the num?